Completed preliminary parse tree to AST mapper

jenrik opened the pull request:
Added terminator to constant declaration.
Named production parts for ease of use.
Collapsed expressions. ANTLR 4 automaticly does precedence based on
declaration order.
Updated mapper to use named production parts.
Added missing modulo expression mapping.
Added missing function expression mapping.
Added state index expression mapping.
Added missing array literal expression mapping.
Added missing if statement mapping.
Added missing return statement mapping.
Added missing neighbourhood declaration mapping.
Added function declaration mapping.
Added syntax stress example.

jenrik committed:
Completed preliminary parse tree to AST mapper

Added terminator to constant declaration.
Named production parts for ease of use.
Added empty statement. Writing a single semicolon adds a no-op
statement.
Collapsed expressions. ANTLR 4 automaticly does precedence based on
declaration order.
Updated mapper to use named production parts.
Added missing modulo expression mapping.
Added missing function expression mapping.
Added state index expression mapping.
Added missing array literal expression mapping.
Added missing if statement mapping.
Added missing return statement mapping.
Added missing neighbourhood declaration mapping.
Added function declaration mapping.
Added syntax stress example.

NicEastvillage commented on compiler/src/main/antlr/Cellmata.g4:
```diff
@@ -93,37 +100,37 @@ bool_literal
+    | OP_NOT value=expr # inverseExpr
+    | value=expr OP_INCREMENT # postIncExpr
+    | value=expr OP_DECREMENT # postDecExpr
+    | value=expr SQ_BRACKET_START DIGITS SQ_BRACKET_END # arrayLookupExpr
```
```suggestion
    | value=expr SQ_BRACKET_START expr SQ_BRACKET_END # arrayLookupExpr
```

NicEastvillage commented on compiler/src/main/kotlin/ast/ast.kt:
```diff
@@ -76,27 +75,22 @@ data class InverseExpr(val ctx: ParseTree, val value: Expr) : Expr()
 data class ArrayLookupExpr(val ctx: ParseTree, val ident: String = MAGIC_UNDEFINED_STRING, val index: Int = -1) : Expr()
 
-data class ArrayBodyExpr(val ctx: ParseTree, val type: String?, val values: List<Expr>) : Expr()
+data class ArrayBodyExpr(val ctx: ParseTree, val type: String? = MAGIC_UNDEFINED_STRING, val values: List<Expr>) : Expr()
```
I think we should make types to an enum.

jenrik replied:
I'm not sure enums are the right solution, how would you represent an array of integers?

NicEastvillage commented on compiler/src/main/kotlin/ast/ast.kt:
```diff
@@ -114,14 +108,23 @@ data class StateDecl(
 ) : Decl()
 
-data class Coordinate(val axes: List<Int>)
+data class Coordinate(val ctx: ParseTree, val axes: List<Int> = emptyList())
```
```suggestion
data class Coordinate(val ctx: ParseTree, val coordX: Int, val coordY: Int?)
```

jenrik replied:
Having the axes as a list allows for easier data manipulation and validation, and have proven useful to me

NicEastvillage commented on compiler/src/main/kotlin/ast/mapper.kt:
```diff
@@ -109,34 +114,55 @@ private fun visitExpr(node: ParseTree): Expr {
+                // null???
+                is CellmataParser.If_stmt_elseContext -> visitCodeBlock(node.if_stmt_else().code_block())
+                else -> throw AssertionError("Unexpected tree node")
+            }
```
Here will be problems when the else-case is not present as the comment `// nul???` also suggests

NicEastvillage commented on compiler/src/main/kotlin/ast/mapper.kt:
```diff
@@ -146,38 +172,37 @@ private fun visitDecl(node: ParseTree): Decl {
+        .filter { !(it is CellmataParser.StmtContext && it.childCount == 0) }
+        .filter { !(it is CellmataParser.StmtContext && it.getChild(0) is TerminalNode ) }
+        .map { visitStmt(it) }
+}
```
This is weird. What is this filtering for? Why isn't children just 0 or more statements?

NicEastvillage commented on compiler/src/main/kotlin/main.kt:
```diff
@@ -22,4 +16,6 @@ fun main() {
 
     println(ast)
+
+    listOf<Unit>()
```
Why??

NicEastvillage reviewed (changes requested):
A few things seems a bit weird. Check those please.

jenrik committed:
Resolved NicEastVillage's comments, and improved mapper

Added name to more production parts.
Changed arrayLookupExpr to use a expression as array index.
Changed AST nodes to used stronger types for the ctx variables.
Improved traversal of literals.
Changed increment and decrement statements to takes arbitrary
expressions instead of variable indetifier.
Changed become to take a expression instead of a fixed state.
Removed unneccesary filtering in visitCodeBlock.
Removed the ability to use a semicolon as an empty statement.

jenrik commented:
The ability to use a semicolon as a no-op statement has been removed

jenrik committed:
Removed dangling list creation in main.kt
